Fails to bootstrap with GCC format hardening

I need to disable the following build hardening setting on my distro to build OW. It's not a huge problem but I wanted to inform you about it.

https://nixos.org/manual/nixpkgs/stable/#format

> Adds the `-Wformat -Wformat-security -Werror=format-security` compiler options. At present, this warns about calls to `printf` and `scanf` functions where the format string is not a string literal and there are no format arguments, as in `printf(foo);`. This may be a security hole if the format string came from untrusted input and contains `%n`.

<details>
  <summary>This is a compile log where the build fails with format hardening enabled:</summary>

```
**** Building the wasm bootstrap
======================= 10:08:36 /build/source/bld/wasm =======================
================== 10:08:36 /build/source/bld/wasm/binbuild ===================
gcc -pipe -D_BLDVER=1300 -D_CYEAR=2021 -fno-asm -fno-common -fsigned-char   -I"/build/source/bld/watcom/h" /build/source/bld/wasm/mkopcod1.c -D_STANDALONE_ -I. -I"/build/source/bld/wasm/h" -D__UNIX__ -D__FLAT__ -D__LINUX__ -D_M_X64 -D__LINUX_X64__ -Wall -Wno-switch -Wno-missing-braces -Wno-parentheses -O -g -o ./mkopcod1.exe  -Wl,-Map,mkopcod1.map /build/source/bld/watcom/binbuild/clibext.lib
./mkopcod1.exe /build/source/bld/wasm/fullops.tok /build/source/bld/wasm/inlnops.tok asmops.gh
gcc -pipe -D_BLDVER=1300 -D_CYEAR=2021 -fno-asm -fno-common -fsigned-char   -I"/build/source/bld/watcom/h" /build/source/bld/wasm/mkopcode.c -D_STANDALONE_ -I. -I"/build/source/bld/wasm/h" -D__UNIX__ -D__FLAT__ -D__LINUX__ -D_M_X64 -D__LINUX_X64__ -Wall -Wno-switch -Wno-missing-braces -Wno-parentheses -O -g -o ./mkopcode.exe  -Wl,-Map,mkopcode.map /build/source/bld/watcom/binbuild/clibext.lib
./mkopcode.exe /build/source/bld/wasm/fullops.tok /build/source/bld/wasm/inlnops.tok asmopsd.gh
optencod -q -rc MSG_USAGE_BASE  ../h/options.gml . . usage.gh linux
awk -v base=MSG_WOMP_BASE -f /build/source/build/makemsg.awk -v OUTFILE=wompmsg.gh ../h/womp.msg
awk -v base=MSG_WASM_BASE -f /build/source/build/makemsg.awk -v OUTFILE=wasmsmsg.gh ../h/wasms.msg
awk -v base=MSG_SHARE_BASE -f /build/source/build/makemsg.awk -v OUTFILE=wasmcmsg.gh /build/source/bld/wasm/h/wasmc.msg
gcc -pipe -c -D_BLDVER=1300 -D_CYEAR=2021 -fno-asm -fno-common -fsigned-char -Wall -Wno-switch -Wno-missing-braces -Wno-parentheses  -O -g  -o main.obj  -DNDEBUG -DBOOTSTRAP -D_M_X64 -D__LINUX_X64__ -D__UNIX__ -D__FLAT__ -D__LINUX__ -D_STANDALONE_       -I. -I"../h" -I"/build/source/bld/womp/h" -I"/build/source/bld/fe_misc/h"  -I"/build/source/bld/trmem" -I"/build/source/bld/watcom/h" ../c/main.c
../c/main.c: In function 'ReadIndirectFile':
../c/main.c:793:9: warning: ignoring return value of 'fread' declared with attribute 'warn_unused_result' [-Wunused-result]
  793 |         fread( env, 1, len, fp );
      |         ^~~~~~~~~~~~~~~~~~~~~~~~
gcc -pipe -c -D_BLDVER=1300 -D_CYEAR=2021 -fno-asm -fno-common -fsigned-char -Wall -Wno-switch -Wno-missing-braces -Wno-parentheses  -O -g  -o write.obj  -DNDEBUG -DBOOTSTRAP -D_M_X64 -D__LINUX_X64__ -D__UNIX__ -D__FLAT__ -D__LINUX__ -D_STANDALONE_       -I. -I"../h" -I"/build/source/bld/womp/h" -I"/build/source/bld/fe_misc/h"  -I"/build/source/bld/trmem" -I"/build/source/bld/watcom/h" ../c/write.c
gcc -pipe -c -D_BLDVER=1300 -D_CYEAR=2021 -fno-asm -fno-common -fsigned-char -Wall -Wno-switch -Wno-missing-braces -Wno-parentheses  -O -g  -o fatal.obj  -DNDEBUG -DBOOTSTRAP -D_M_X64 -D__LINUX_X64__ -D__UNIX__ -D__FLAT__ -D__LINUX__ -D_STANDALONE_       -I. -I"../h" -I"/build/source/bld/womp/h" -I"/build/source/bld/fe_misc/h"  -I"/build/source/bld/trmem" -I"/build/source/bld/watcom/h" ../c/fatal.c
gcc -pipe -c -D_BLDVER=1300 -D_CYEAR=2021 -fno-asm -fno-common -fsigned-char -Wall -Wno-switch -Wno-missing-braces -Wno-parentheses  -O -g  -o womputil.obj  -DNDEBUG -DBOOTSTRAP -D_M_X64 -D__LINUX_X64__ -D__UNIX__ -D__FLAT__ -D__LINUX__ -D_STANDALONE_       -I. -I"../h" -I"/build/source/bld/womp/h" -I"/build/source/bld/fe_misc/h"  -I"/build/source/bld/trmem" -I"/build/source/bld/watcom/h" ../c/womputil.c
gcc -pipe -c -D_BLDVER=1300 -D_CYEAR=2021 -fno-asm -fno-common -fsigned-char -Wall -Wno-switch -Wno-missing-braces -Wno-parentheses  -O -g  -o direct.obj  -DNDEBUG -DBOOTSTRAP -D_M_X64 -D__LINUX_X64__ -D__UNIX__ -D__FLAT__ -D__LINUX__ -D_STANDALONE_       -I. -I"../h" -I"/build/source/bld/womp/h" -I"/build/source/bld/fe_misc/h"  -I"/build/source/bld/trmem" -I"/build/source/bld/watcom/h" ../c/direct.c
../c/direct.c: In function 'get_watcom_argument_string':
../c/direct.c:388:13: error: format not a string literal and no format arguments [-Werror=format-security]
  388 |             sprintf( buffer, regs[A_BYTE][parm] );
      |             ^~~~~~~
../c/direct.c:391:13: error: format not a string literal and no format arguments [-Werror=format-security]
  391 |             sprintf( buffer, regs[A_WORD][parm] );
      |             ^~~~~~~
../c/direct.c:395:17: error: format not a string literal and no format arguments [-Werror=format-security]
  395 |                 sprintf( buffer, regs[A_DWORD][parm] );
      |                 ^~~~~~~
../c/direct.c: In function 'Ret':
../c/direct.c:4042:34: warning: '%lu' directive writing between 1 and 20 bytes into a region of size 15 [-Wformat-overflow=]
 4042 |                     sprintf( p, "%lu", info->parasize );
      |                                  ^~~
../c/direct.c:4042:33: note: directive argument in the range [1, 18446744073709551615]
 4042 |                     sprintf( p, "%lu", info->parasize );
      |                                 ^~~~~
In file included from /nix/store/vzqvzmlalpvanw7f7p79fvk92b6lzi87-glibc-2.32-46-dev/include/stdio.h:866,
                 from ../h/asmglob.h:36,
                 from ../c/direct.c:33:
/nix/store/vzqvzmlalpvanw7f7p79fvk92b6lzi87-glibc-2.32-46-dev/include/bits/stdio2.h:38:10: note: '__builtin___sprintf_chk' output between 2 and 21 bytes into a destination of size 15
   38 |   return __builtin___sprintf_chk (__s, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   39 |       __bos (__s), __fmt, __va_arg_pack ());
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../c/direct.c:4030:34: warning: '%lu' directive writing between 1 and 20 bytes into a region of size 15 [-Wformat-overflow=]
 4030 |                     sprintf( p, "%lu", info->parasize );
      |                                  ^~~
../c/direct.c:4030:33: note: directive argument in the range [1, 18446744073709551615]
 4030 |                     sprintf( p, "%lu", info->parasize );
      |                                 ^~~~~
In file included from /nix/store/vzqvzmlalpvanw7f7p79fvk92b6lzi87-glibc-2.32-46-dev/include/stdio.h:866,
                 from ../h/asmglob.h:36,
                 from ../c/direct.c:33:
/nix/store/vzqvzmlalpvanw7f7p79fvk92b6lzi87-glibc-2.32-46-dev/include/bits/stdio2.h:38:10: note: '__builtin___sprintf_chk' output between 2 and 21 bytes into a destination of size 15
   38 |   return __builtin___sprintf_chk (__s, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   39 |       __bos (__s), __fmt, __va_arg_pack ());
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../c/direct.c:4035:34: warning: '%lu' directive writing between 1 and 20 bytes into a region of size 15 [-Wformat-overflow=]
 4035 |                     sprintf( p, "%lu", info->parasize );
      |                                  ^~~
../c/direct.c:4035:33: note: directive argument in the range [1, 18446744073709551615]
 4035 |                     sprintf( p, "%lu", info->parasize );
      |                                 ^~~~~
In file included from /nix/store/vzqvzmlalpvanw7f7p79fvk92b6lzi87-glibc-2.32-46-dev/include/stdio.h:866,
                 from ../h/asmglob.h:36,
                 from ../c/direct.c:33:
/nix/store/vzqvzmlalpvanw7f7p79fvk92b6lzi87-glibc-2.32-46-dev/include/bits/stdio2.h:38:10: note: '__builtin___sprintf_chk' output between 2 and 21 bytes into a destination of size 15
   38 |   return __builtin___sprintf_chk (__s, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   39 |       __bos (__s), __fmt, __va_arg_pack ());
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: some warnings being treated as errors
Error(E42): Last command making (direct.obj) returned a bad status
Error(E02): Make execution terminated
<wmake -h -f ../binmake bootstrap=1> => non-zero return: 512
Build failed
builder bootstrap build error
```

</details>

Systems are 64-bit NixOS 20.09 & Ubuntu 18.04.5 with Nixpkgs build system.

--- a/bld/gui/ui/c/guildstr.c
+++ b/bld/gui/ui/c/guildstr.c
@@ -65,7 +65,7 @@ bool GUILoadStrInit( const char *fname )
     }
     CloseResFile( &hInstance );
     printf( NO_RES_MESSAGE_PREFIX );
-    printf( fname );
+    printf( "%s", fname );
     printf( NO_RES_MESSAGE_SUFFIX );
     return( false );
 }

--- a/bld/wasm/c/direct.c
+++ b/bld/wasm/c/direct.c
@@ -385,14 +385,14 @@ static bool get_watcom_argument_string( char *buffer, int size, int *parm_number
     } else {
         switch( size ) {
         case 1:
-            sprintf( buffer, regs[A_BYTE][parm] );
+            sprintf( buffer, "%s", regs[A_BYTE][parm] );
             break;
         case 2:
-            sprintf( buffer, regs[A_WORD][parm] );
+            sprintf( buffer, "%s", regs[A_WORD][parm] );
             break;
         case 4:
             if( Use32 ) {
-                sprintf( buffer, regs[A_DWORD][parm] );
+                sprintf( buffer, "%s", regs[A_DWORD][parm] );
             } else {
                 if( parm > 2 ) {
                     *on_stack = true;

--- a/bld/wl/fcenable/c/fcenable.c
+++ b/bld/wl/fcenable/c/fcenable.c
@@ -569,7 +569,7 @@ int main(int argc, char **argv )
     ClassList->lnameidx = 0;
     memcpy( ClassList->name, DEF_CLASS, sizeof( DEF_CLASS ) - 1 );
     if( ( argc < 2 ) || ( argv[1][0] == '?' ) ) {
-        printf( HelpMsg );
+        printf( "%s", HelpMsg );
     } else {
         argv++;     // skip the program name
         retval = Spawn1( ProcessFiles, argv );

